compiler/rustc_parse/src/parser/nonterminal.rs

```diff
@@ -169,7 +170,19 @@ impl<'a> Parser<'a> {
                     return Err(self.struct_span_err(self.token.span, msg));
                 }
             }
-        })
+        };
+
+        // If tokens are supported at all, they should be collected.
+        if !matches!(nt.tokens_mut(), None | Some(Some(..))) {
```

Comment:
I think this could be written more clearly as `matches!(nt.tokens_mut(), Some(None)`